File: wapp/src/process/session/queue.lua
-- Simple queue implementation for managing task payloads
local queue = {}
queue.__index = queue

-- Create a new queue
function queue.new()
    local self = setmetatable({}, queue)
    self.items = {}
    self.head = 1
    self.tail = 0
    self.size = 0
    return self
end

-- Add an item to the queue
function queue:enqueue(item)
    self.tail = self.tail + 1
    self.items[self.tail] = item
    self.size = self.size + 1
    return true
end

-- Remove and return the next item from the queue
function queue:dequeue()
    if self.size <= 0 then
        return nil, "Queue is empty"
    end

    local item = self.items[self.head]
    self.items[self.head] = nil
    self.head = self.head + 1
    self.size = self.size - 1

    -- Reset indices if queue is empty to avoid growing indices indefinitely
    if self.size == 0 then
        self.head = 1
        self.tail = 0
    end
    return item
end

-- Check if the queue is empty
function queue:is_empty()
    return self.size == 0
end

-- Get the current size of the queue
function queue:get_size()
    return self.size
end

-- Clear all items from the queue
function queue:clear()
    self.items = {}
    self.head = 1
    self.tail = 0
    self.size = 0
end

-- Peek at the next item without removing it
function queue:peek()
    if self.size <= 0 then
        return nil
    end
    return self.items[self.head]
end

return queue

File: system/scheduler/actor/queue.go
// SPDX-License-Identifier: MPL-2.0

package actor

import (
	"sync"
	"sync/atomic"
)

// Queue is a thread-safe FIFO queue using a ring buffer.
//
// Design choices:
//   - Mutex-based for simplicity (global queue is not the hot path)
//   - Ring buffer to avoid slice append allocations
//   - Grows on demand but never shrinks (avoids churn)
//   - Atomic count for fast empty checks without locking
//
// Usage:
//   - New process submissions
//   - Cross-worker transfers after async completion
//   - Initial work distribution
//
// The global queue is NOT the hot path. Workers primarily use local deques.
// Global queue is only accessed for:
//  1. Submit() - new process enters system
//  2. Complete() from async handler - process re-enters from outside worker
//  3. findWork() fallback - when local deque is empty
type Queue struct {
	items     []*Processor
	head      int // Read position (oldest item)
	tail      int // Write position (next empty slot)
	count     int
	atomicLen atomic.Int32 // Fast lock-free length check
	mu        sync.Mutex
}

// NewQueue creates a queue with initial capacity.
// Capacity should be sized for expected concurrent processes.
func NewQueue(capacity int) *Queue {
	if capacity < 16 {
		capacity = 16
	}
	return &Queue{
		items: make([]*Processor, capacity),
	}
}

// Push adds a processor to the queue (thread-safe).
// Grows buffer if needed.
func (q *Queue) Push(p *Processor) {
	q.mu.Lock()

	if q.count == len(q.items) {
		q.grow()
	}

	q.items[q.tail] = p
	q.tail = (q.tail + 1) % len(q.items)
	q.count++
	q.atomicLen.Add(1)

	q.mu.Unlock()
}

// Pop removes and returns the oldest processor (FIFO).
// Returns nil if queue is empty.
func (q *Queue) Pop() *Processor {
	// Fast path: lock-free empty check
	if q.atomicLen.Load() == 0 {
		return nil
	}

	q.mu.Lock()

	if q.count == 0 {
		q.mu.Unlock()
		return nil
	}

	p := q.items[q.head]
	q.items[q.head] = nil // Clear reference for GC
	q.head = (q.head + 1) % len(q.items)
	q.count--
	q.atomicLen.Add(-1)

	q.mu.Unlock()
	return p
}

// PopN removes and returns up to n processors.
// Returns actual count taken. Useful for batch transfers.
func (q *Queue) PopN(dst []*Processor) int {
	// Fast path: lock-free empty check
	if q.atomicLen.Load() == 0 {
		return 0
	}

	q.mu.Lock()

	n := q.count
	if n > len(dst) {
		n = len(dst)
	}

	for i := 0; i < n; i++ {
		dst[i] = q.items[q.head]
		q.items[q.head] = nil // Clear reference for GC
		q.head = (q.head + 1) % len(q.items)
	}
	q.count -= n
	q.atomicLen.Add(int32(-n))

	q.mu.Unlock()
	return n
}

// Len returns current item count (lock-free).
func (q *Queue) Len() int {
	return int(q.atomicLen.Load())
}

// IsEmpty returns true if queue has no items (lock-free).
func (q *Queue) IsEmpty() bool {
	return q.atomicLen.Load() == 0
}

// grow doubles the buffer capacity.
// Must be called while holding the lock.
func (q *Queue) grow() {
	newCap := len(q.items) * 2
	newItems := make([]*Processor, newCap)

	// Copy items maintaining FIFO order
	for i := 0; i < q.count; i++ {
		newItems[i] = q.items[(q.head+i)%len(q.items)]
	}

	q.items = newItems
	q.head = 0
	q.tail = q.count
}
